### ISSUE TITLE
Python 2.7 not available on Heroku-18

### ISSUE THREAD

**Issue by edmorley** [heroku/heroku-buildpack-python#735]
Heroku-18 is now [out of beta](https://devcenter.heroku.com/changelog-items/1453), however using `python-2.7.15` in `runtime.txt` does not work:

```
-----> Python app detected
2018-07-25 10:03:35 -----> Stack has changed from heroku-16 to heroku-18, clearing cache
2018-07-25 10:03:36 -----> Installing python-2.7.15
2018-07-25 10:03:36  !     Requested runtime (python-2.7.15) is not available for this stack (heroku-18).
2018-07-25 10:03:36  !     Aborting.  More info: https://devcenter.heroku.com/articles/python-support
```

Only the Python 3 archives are available:

```
$ aws s3 ls --no-sign-request --recursive s3://lang-python/heroku-18/
2018-05-04 20:20:47    3008419 heroku-18/libraries/sqlite.tar.gz
2018-04-30 17:49:09         66 heroku-18/libraries/vendor/libffi.tar.gz
2018-05-22 20:32:34   52974289 heroku-18/runtimes/python-3.6.0.tar.gz
2018-05-22 20:50:10   53123812 heroku-18/runtimes/python-3.6.1.tar.gz
2018-05-22 21:40:50   38202978 heroku-18/runtimes/python-3.6.2.tar.gz
2018-05-22 22:22:16   38339163 heroku-18/runtimes/python-3.6.3.tar.gz
2018-05-04 20:55:28   38363897 heroku-18/runtimes/python-3.6.4.tar.gz
2018-05-04 22:33:00   38670586 heroku-18/runtimes/python-3.6.5.tar.gz
2018-06-28 16:16:16   36281492 heroku-18/runtimes/python-3.6.6.tar.gz
2018-06-28 16:20:34   38149401 heroku-18/runtimes/python-3.7.0.tar.gz
```

This is mentioned here:
https://devcenter.heroku.com/articles/python-support#supported-runtimes

...and so presumably intentional. I guess the reasoning is that Heroku 18 is supported until 2023, beyond the lifetime of Python 2. However I don't see that that's much different from eg older Python point releases still being available at that point?

We have no plans to be using Python 2 beyond it's EOL (and should be off it well before then thankfully), and it would be ideal if the stack update wasn't blocked on our move to Python 3.

Please could someone generate/upload the archive for Python 2.7.15 for Heroku-18?

**Comment by sigmavirus24** [comment]
This is an intentional decision to not support for Python 2.7 on Heroku-18. We have no plans to support Python 2.7 on Heroku-18.

I'm glad you have no intention of using it past its End of Life, but not all customers will be willing to commit to the effort involved to migrating to Python 3 at that point. Given that 2020 is the hard end of life for Python 2.7 and security vulnerabilities will be found for the 3 years following that, we can not in good conscience encourage users to use what will become out-dated and insecure software past that point. Already, major software packages that are critical path for the ecosystem are dropping Python 2.7 support or introducing plans to do so prior to 2020. This means that security vulnerabilities in those packages will force Python 2.7 into a de facto insecure posture.

**Comment by edmorley** [comment]
Thank you for the reply - makes sense. 

This ticket will at least give me another thing to point at in addition to Django 2.x/gunicorn 20/... when making the case for prioritising the remaining Python 3 tickets for the project in question (which until now has been something I've had to chip away at in my spare time) - and at least I can now say I've asked ;-)

**Comment by kennethreitz** [comment]
✨🍰✨

**Comment by kennethreitz** [comment]
P.S. This was @dzuelke's idea :)

**Comment by sigmavirus24** [comment]
A very good idea to say the least

**Comment by dzuelke** [comment]
Maybe for some additional context, in the interest of enshrining the reasoning into the eternal memory of GitHub Issues... :)

The rationale is that heroku-16 will be supported until 2021, beyond the EOL of Python 2.7, which means people who want to use Python 2.7 until its EOL date will be able to do so on a supported stack.

The heroku-18 introduction was simply a good opportunity to at least check people's reactions; the idea being that it's a lot easier to add Python 2.7 if there is enough demand, than to take it away or somehow phase it out later ;)

It will be interesting to see what happens once heroku-18 becomes the default stack though...

**Comment by edmorley** [comment]
Since people are linking to this old issue - note that since it was closed, Python 2.7 has now been made available on Heroku-16/18 (#745).
